cmd/routesrv/main.go

```diff
@@ -0,0 +1,262 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/cmd/routesrv/options"
+	"github.com/zalando/skipper/config"
+	"github.com/zalando/skipper/dataclients/kubernetes"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/routing"
+	"github.com/zalando/skipper/tracing"
+)
+
+// eskipBytes keeps eskip-formatted routes as a byte slice and
+// provides synchronized r/w access to them. Additionally it can
+// serve as an HTTP handler exposing its content.
+type eskipBytes struct {
+	data        []byte
+	mu          sync.RWMutex
+	initialized bool
+
+	tracer ot.Tracer
+}
+
+func (e *eskipBytes) bytes() ([]byte, bool) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	return e.data, e.initialized
+}
+
+func (e *eskipBytes) formatAndSet(routes []*eskip.Route) (int, bool) {
+	buf := &bytes.Buffer{}
+	eskip.Fprint(buf, eskip.PrettyPrintInfo{}, routes...)
+
+	e.mu.Lock()
+	oldData := e.data
+	e.data = buf.Bytes()
+	e.initialized = true
+	e.mu.Unlock()
+
+	return len(oldData), oldData == nil
+}
+
+func (e *eskipBytes) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	span := tracing.CreateSpan("serve_routes", context.TODO(), e.tracer)
+	defer span.Finish()
+
+	if data, initialized := e.bytes(); initialized {
+		w.Write(data)
+	} else {
+		w.WriteHeader(http.StatusNotFound)
+	}
+}
+
+// eskipBytesStatus serves as an HTTP health check for the referenced eskipBytes.
+// Reports healthy only when the bytes were initialized (set at least once).
+type eskipBytesStatus struct {
+	b *eskipBytes
+}
+
+var msgRoutesNotInitialized = []byte("routes were not initialized yet")
+
+func (s *eskipBytesStatus) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if _, initialized := s.b.bytes(); initialized {
+		w.WriteHeader(http.StatusOK)
+	} else {
+		w.WriteHeader(http.StatusServiceUnavailable)
+		w.Write(msgRoutesNotInitialized)
+	}
+}
+
+type pollerMetrics struct {
+	startedTimestamp     prometheus.Gauge
+	initializedTimestamp prometheus.Gauge
+	updatedTimestamp     prometheus.Gauge
+}
+
+func newPollerMetrics() *pollerMetrics {
+	return &pollerMetrics{
+		startedTimestamp: promauto.NewGauge(prometheus.GaugeOpts{
+			Namespace: "routesrv",
+			Subsystem: "poller",
+			Name:      "started_timestamp",
+		}),
+		initializedTimestamp: promauto.NewGauge(prometheus.GaugeOpts{
+			Namespace: "routesrv",
+			Subsystem: "poller",
+			Name:      "initialized_timestamp",
+		}),
+		updatedTimestamp: promauto.NewGauge(prometheus.GaugeOpts{
+			Namespace: "routesrv",
+			Subsystem: "poller",
+			Name:      "updated_timestamp",
+		}),
```

Comment:
I am struggling to grasp what do we measure here

Reply:
Added prom native help messages to them.